File: common/src/main/clojure/com/github/lukebemish/clojurewrapper/api/platform.clj
(ns com.github.lukebemish.clojurewrapper.api.platform
  (:require [clojure.string :as string])
  (:import (com.github.lukebemish.clojurewrapper.env Platform)
           (com.github.lukebemish.clojurewrapper.util RunPlatform)
           (clojure.lang Symbol)))

(def forge Platform/FORGE)
(def fabric Platform/FABRIC)

(defn get-platform [] (Platform/getPlatform))
(defn unimpl-error [] (Platform/unimplError))

(defn fabric? [] (= (get-platform) fabric))
(defn forge? [] (= (get-platform) forge))

(def fn-platform (memoize (fn [symb] (RunPlatform/runPlatform (str symb)))))
(defn run-platform [^Symbol symb & context]
  (let [found (fn-platform symb)]
    (if (nil? found)
      (unimpl-error)
      (apply found context))))

File: common/src/main/java/com/github/lukebemish/clojurewrapper/env/Platform.java
package com.github.lukebemish.clojurewrapper.env;

import dev.architectury.injectables.annotations.ExpectPlatform;

public enum Platform {
    FORGE("forge"),
    FABRIC("fabric");

    private final String name;

    @ExpectPlatform
    public static Platform getPlatform() {
        throw new AssertionError();
    }

    public static Object unimplError() {
        throw new AssertionError();
    }

    Platform(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }
}
